cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,382 @@
+            args.append(f'qubit_map={dict_str(self.qubit_map)}')
+        if self.measurement_key_map != base_op.measurement_key_map:
+            args.append(f'key_map={dict_str(self.measurement_key_map)}')
+        if self.param_resolver != base_op.param_resolver:
+            args.append(f'params={self.param_resolver.param_dict}')
+        if self.repetitions != 1:
+            args.append(f'loops={self.repetitions}')
+        if not args:
+            return f'{header}\n{circuit_msg}'
+        return f'{header}\n{circuit_msg}({", ".join(args)})'
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': 'CircuitOperation',
+            'circuit': self.circuit,
+            'repetitions': self.repetitions,
+            # JSON requires mappings to have keys of basic types.
+            # Pairs must be sorted to ensure consistent serialization.
+            'qubit_map': sorted(self.qubit_map.items()),
+            'measurement_key_map': self.measurement_key_map,
+            'param_resolver': self.param_resolver,
+        }
+
+    @classmethod
+    def _from_json_dict_(
+        cls, circuit, repetitions, qubit_map, measurement_key_map, param_resolver, **kwargs
+    ):
+        result = (
+            cls(circuit)
+            .with_qubit_mapping(dict(qubit_map))
+            .with_measurement_key_mapping(measurement_key_map)
+            .with_params(param_resolver)
+        )
+        if repetitions != 1:
+            return result.repeat(repetitions)
+        return result
+
+    # Methods for constructing a similar object with one field modified.
+
+    def with_circuit(self, new_circuit: 'cirq.AbstractCircuit') -> 'CircuitOperation':
+        """Returns a copy of this operation with the provided circuit.
+
+        Args:
+            new_circuit: The circuit to pack into a copy of this operation.
+                Key mappings, parameter values, and repetitions are preserved.
+
+        Returns:
+            A copy of this operation wrapping the provided circuit.
+
+        Raises:
+            NotImplementedError: The new circuit contains measurements, but
+                this operation has repetitions.
+        """
+        if protocols.is_measurement(new_circuit) and self._repetitions != 1:
+            raise NotImplementedError('Measurements cannot be added to looped circuits.')
+        new_op = self.copy()
+        new_op._circuit = new_circuit.freeze()
+        return new_op
+
+    def repeat(
+        self,
+        repetitions: INT_TYPE,
+        modulus: INT_TYPE = 0,
+        allow_invert: bool = False,
+        validate_modulus: bool = False,
+    ) -> 'CircuitOperation':
+        """Returns a copy of this operation repeated 'repetitions' times.
+
+        Args:
+            repetitions: Number of times this operation should repeat. This
+                is multiplied with any pre-existing repetitions.
+            modulus: If this is specified, the final repetition count will be
```

Comment:
Why do we need `modulus` + `allow_invert`? Do you have a motivating use case?

Reply:
Adam highlights this as a useful feature in the [RFC](https://tinyurl.com/cirq-circuitoperation-rfc) - feel free to check with him for specific use cases.

Reply:
Ah I see, cool, inviting @viathor to comment.

Reply:
I imagined it a little differently: at the instantiation time of a CircuitOperation the user could provide an optional modulus for exponentiation as a hint (and we could validate it in `__init__` the way it is done in this method).

Regarding a use-case: I don't have one that I need at the moment. The rationale for this was that in many situations one wants to scan over a range of exponents. This range is often bounded, but it may not be clear what the bound for a particular circuit is. So if the user wants to scan the full range, they have to figure out what "full" means. Of course, they also have to do this in order to provide the hint, but the code that prepares a CircuitOperation and the code that consumes it may be different and this is easier to do for the producer of CircuitOperation.

That said, at the time I wrote the comment I was trying to anticipate a need and wanted to see whether others find it useful. If there is currently no explicit need for it, we can drop it and add it when users ask for it.

Reply:
I see, thank you! @95-martin-orion what do you think? I think we can keep it, but moving modulus with validation to the constructor makes more sense to me, as then the exponentiation code could make use of the modulus being "stored" as well.

Reply:
Comment from @maffoo on #3509 (a prior version of this change) regarding modulus in the constructor:

> I'm not sure this is a good idea. When we ask hardware to repeat a circuit we typically want it to run that many times, we don't want things that are equivalent to identities to get optimized away (think about randomized benchmarking). Such transformations could be implemented in an optimization pass, but it seems odd to bake this in to the gate itself.

Reply:
Well, for randomized benchmarking we wouldn't set the modulus. Having to keep repeating the modulus puts the extra effort on the algorithm developer's side: in case the algo developer wants to use a "modulus optimized subcircuit" across a range of different exponents, they have to ensure that for all the exponents the validation will happen - alternatively, run the validation once and then use it - but they have to ensure then to use the same modulus across all calls, i.e. they'll have to design to store the modulus somewhere. With the constructor based method, they can choose and validate once and reuse and exponentiate the subcircuit freely with built-in modulus optimization.
@maffoo what do you think?

Reply:
Discussed out-of-band: removing `modulus` from this PR and opening an issue to track it as a separate feature.

Reply:
This change has been applied.